### ISSUE TITLE
A SHP file makes many plugins crash

### ISSUE THREAD

**Issue by donkeychen** [qgis/QGIS#40084]
hi, I added a point vector shp file, the result of "check validity", all 666 features are valid.
But, when I analyzed it by plugins(something like topology checker)
it made the qgis crashed (every time, including QGIS 2.x and QGIS 3.x)

You can download the data from
https://drive.google.com/file/d/16B_72DbB6765cEh6MLCYmBVXDvd3dFAN/view?usp=sharing

(by export/add geometry columns, and save as csv, than add this point layer from csv, the crash issue is gone.
this makes me confused.
Is any data error in the SHP file? How they passed the validity check? 
Thank you.

**QGIS Info**
QGIS Version: 3.12.2-Bucure?ti
QGIS code revision: 8a1fb33634
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 3.0.4
Running against GDAL: 3.0.4

**System Info**
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.18363

**Comment by gioman** [comment]
>  https://drive.google.com/file/d/16B_72DbB6765cEh6MLCYmBVXDvd3dFAN/view?usp=sharing

@donkeychen link is not public

>  QGIS Version: 3.12.2-Bucure?ti

too old, you must use the latest 3.10 (LTR) or the latest 3.16. Also please try with a new QGIS profile, no 3rd party plugins installed.

**Comment by donkeychen** [comment]
@gioman 
Thank for your reminding, I changed the sharing setting, please try the file again.
(the encoding is "Big5")

I saw the Long term release repository (most stable) is 3.10. :(
I tried the newest 3.16 for testing anyway.

Add the vector layer 
1. check validity => # of Valid output: 666, # of Invalid output: 0
2. topology checker => topology rule settings: must not have duplicate => run
QGIS crashed after a long time non responsing

Stack Trace


Index::~Index :
qsort :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
Index::~Index :
QgsSpatialIndex::addFeature qgsspatialindex.cpp:453
QgsSpatialIndex::addFeature qgsspatialindex.cpp:410
topolTest::createIndex topoltest.cpp:1360
topolTest::runTest topoltest.cpp:1436
checkDock::runTests checkdock.cpp:358
checkDock::validate checkdock.cpp:397
QMetaObject::activate :
QAction::activate :
QAbstractButton::click :
QAbstractButton::mouseReleaseEvent :
QToolButton::mouseReleaseEvent :
QWidget::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify qgsapplication.cpp:458
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify qgsapplication.cpp:458
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
UserCallWinProcCheckWow :
DispatchMessageWorker :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main main.cpp:1637
BaseThreadInitThunk :
RtlUserThreadStart :




QGIS Info
QGIS Version: 3.16.0-Hannover
QGIS code revision: 43b64b13f3
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 3.1.4
Running against GDAL: 3.1.4



System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.19042

**Comment by gioman** [comment]
>     2\. topology checker => topology rule settings: must not have duplicate => run
>         QGIS crashed after a long time non responsing

Crash confirmed also on Ubuntu 20.04 and QGIS 3.16

**Comment by donkeychen** [comment]
information update
By checking the following processes
1. add the shp file. => **(layer1)**
2. set the CRS to TWD97/TM2 zone 121 (EPSG:3826)
3. run [vector]->[geometry tools] -> [add geometry attributes] for layer1 => **(layer2)**
4. export the result of step 3 (layer2), save the feature as "csv" file (same CRS, and same encoding (big5))
5. add layer from delimited text, file format: CSV; geometry definition: point, X field: xcoord, Y field: ycoord, CRS: EPSG:3826 => **(layer3)** 

-----------------
There are 3 layers(in steps 1.3.5 above respectively) in the layers panel now,
by using same topology checker for these layers respectively, 
result: 
**(layer3)**: 19 errors found
**(layer2)**: crashed
**(layer1)**: crashed
-----------------
I am confused that the processing of getting x, y coordinates from **(layer1)** is successful, and what difference between **(layer2)** and **(layer3)** (with same attribute table)

(this issue can be duplicated in QGIS 2, I understand that version 2.x not update anymore, for you reference)
thank you.

**Comment by gioman** [comment]
@donkeychen geometry checker and topology checker have different code base, they were developed separately at different times independently. They really should be merged (there is a feature request for it) and return coherent results.

**Comment by donkeychen** [comment]
hi @gioman , do you mean the functions in the vector menu is **"geometry checker"**? (plugins => processing)
================
Processing
Spatial data processing framework for QGIS

Category	Analysis
More info	homepage   bug tracker   code repository
Author	Victor Olaya
Installed version	2.12.99
================
and the topology checker is another one
================
Topology Checker
A Plugin for finding topological errors in vector layers

Category	Vector
Installed version	Version 0.1
================
I don't know if the original shape file (exported from ArcGIS) is corrupted, and how to check it.

I saved the result of step 5 as another SHP file. (original point shape file -> csv with coordinate -> new shape file)
[20201118.zip](https://github.com/qgis/QGIS/files/5560081/20201118.zip)

This data will not make the topology checker crashed.
><, oops...... knowing too less about ESRI SHP file.

**Comment by gioman** [comment]
@donkeychen as far as I can see if the wrong encoding that breaks the topology checker tool. If the encoding is set to big5 then the tool does not crash qgis. Can you confirm?

**Comment by donkeychen** [comment]
hi, @gioman ,
I recorded my testing process on YT
https://youtu.be/WdGcjyF3eW4
You can check the outline for detail steps.
Thank you.

**Comment by donkeychen** [comment]
I don't know if the wrong encoding affects the checker.
The original one, with correct encoding did make the checker crashed on my WIN10
https://www.youtube.com/watch?v=WdGcjyF3eW4&t=2m22s shows it.

**Comment by gioman** [comment]
> I don't know if the wrong encoding affects the checker.

@donkeychen it is.

**Comment by agiudiceandrea** [comment]
Hi @donkeychen, the provided shapefile contains a "null geometry" not viewable point feature: both x and y coordinates of the feature with Feature ID ($id) = 332 (zero based) contains the double (float 64) minimum value (~ -1.7976931348623157E+308).

You can find it easily: just zoom to the layer, select all the visible feature in the map extent (665 features selected), open the attribute table, invert the selection, show selected features. This feature is responsible for the crash.

If you use "Validate Extent" instead of "Validate All" with the Topology Checker plugin, QGIS will not crash.

**Comment by agiudiceandrea** [comment]
Hi @donkeychen, the provided shapefile contains a "null geometry" not viewable point feature: both x and y coordinates of the feature with Feature ID ($id) = 332 (zero based) [OBJECTID=89163] contains the double (float 64) minimum value (~ -1.7976931348623157E+308).

You can find it easily: just zoom to the layer, select all the visible feature in the map extent (665 features selected), open the attribute table, invert the selection, show selected features. This feature is responsible for the crash.

If you use "Validate Extent" instead of "Validate All" with the Topology Checker plugin, QGIS will not crash.

Anyway, it should not crash also using "Validate All"...

**Comment by gioman** [comment]
> Anyway, it should not crash also using "Validate All"...

@agiudiceandrea thanks for the analysis, funny thing, if I set the correct encoding the topology checker tool does not crash QGIS.

**Comment by donkeychen** [comment]
Hi, @agiudiceandrea,
Thank you for your analysis very much. 
I checked the file again, the feature count of SHP2CSV2SHP was 665!
(ORIGINAL: 666, valid: 666, with geometry column with -1.79769313486232e+308: 666, csv: 666)
I always used "check validity" to find the error geometry out, the error features would be shown in invalid output and the error was "Unknown geometry type 0" (including point/line/polygon layers).

The output of "check validity" (all 666 features valid) is my blind spot.
Coordinates with a negative value (overflow?) is valid. (pondering......) I never considered the possibility!
You taught me a new method to pick the error features out.
This is a precious experience for me.

**Comment by agiudiceandrea** [comment]
> > Anyway, it should not crash also using "Validate All"...
> 
> @agiudiceandrea thanks for the analysis, funny thing, if I set the correct encoding the topology checker tool does not crash QGIS.
@gioman could you please test with [TEST.zip](https://github.com/qgis/QGIS/files/5577498/TEST.zip)?
It is a shapefile with the same identical geometries of the provided one and an attribute table with only ASCII chars (declared as windows-1252/cp1252/Windows ANSI/...) to avoid encoding issues.
QGIS crashes for me with this file too.

My understanding is that QGIS crashes when the Topology Checker tries to add the features to a QgsSpatialIndex in this loop:

https://github.com/qgis/QGIS/blob/eec1695eb87cfc891af0b52162be6b51941d80fb/src/plugins/topology/topolTest.cpp#L1347-L1362

**Comment by gioman** [comment]
> @gioman could you please test with [TEST.zip](https://github.com/qgis/QGIS/files/5577498/TEST.zip)?

@agiudiceandrea yeah... forget my comments about the encoding, now it crashes regardless, it probably was a fluke. I updated the title.

**Comment by uclaros** [comment]
The attached dataset does not contain any Null geometries as they are in general handled well by qgis. The geometry that is not displayed and is causing the crash is just far far away. The point's coordinates (both x and y) are roughly -1.79e+308. That's a pretty big number!

**Comment by agiudiceandrea** [comment]
Hi @uclaros, you are right: it is a "null geometry" only according to ESRI ArcMap 9.3.1 Geometry Check Tool.

Anyway, the issue is not specific to shapefiles, it also happens if you copy all the features to a memory layer and then use the Topology Checker ("must not have duplicate" constraint) with the memory layer.

It could be a libspatialindex issue: see these QGIS issue reports with similar stack traces and behavior https://github.com/qgis/QGIS/issues/29630 and https://github.com/qgis/QGIS/issues/28751 and libspatialindex issue report https://github.com/libspatialindex/libspatialindex/issues/107.

Testing with pyqgis the Topology Checker [createIndex](https://github.com/qgis/QGIS/blob/1b5682579139d37a3ee5b39bbe331ae06a7cee10/src/plugins/topology/topolTest.cpp#L1327-L1365) procedure using the [TEST.zip](https://github.com/qgis/QGIS/files/5577498/TEST.zip) layer

```python
layer = iface.activeLayer()
index = QgsSpatialIndex()
f = QgsFeature()
fit = QgsFeatureIterator( layer.getFeatures( QgsFeatureRequest().setNoAttributes() ) )
while fit.nextFeature( f ):
	print ( str(f.id()) );
	index.addFeature( f );
```
the loop generates always a QGIS crash at a random feature after the f.id() = 332.

The loop doesn't crash if the feature with f.id() = 332 is removed from the layer or is excluded in the loop.

Nevertheless, the spatial index can be created without issue simply using `index = QgsSpatialIndex( layer.getFeatures() )`.

Could it be possible to refactor the createIndex function avoiding adding the features to the index one by one in a loop?

**Comment by uclaros** [comment]
> Could it be possible to refactor the createIndex function avoiding adding the features to the index one by one in a loop?

You wont be able to check if the user pressed cancel...

What I also noticed is that while
```python
layer = iface.activeLayer()
index = QgsSpatialIndex()
for f in layer.getFeatures():
    index.addFeature( f )
```
crashes, the following works:
```python
layer = iface.activeLayer()
index = QgsSpatialIndex()
for f in layer.getFeatures():
    if f.id() != 332
        index.addFeature( f )
index.addFeature( layer.getFeature( 332 ) )
```

**Comment by agiudiceandrea** [comment]
> You wont be able to check if the user pressed cancel...

QgsSpatialIndex has a  QgsFeedback parameter. Could it be used for the same purpose?

**Comment by uclaros** [comment]
> Maybe a similar approach could be used in the Topology Checker plugin.

Indeed, it could! It would also speed-up the index building! But some refactoring is needed to replace the current signals and use a QgsFeedback object.
Also, this will not fix the actual problem which is handling of huge float coords, only avoid it.

**Comment by Pedro-Murteira** [comment]
I can still reproduce a crash with the original dataset. Topology Checker > Must not have duplicates > QGIS Crashes. (QGIS 3.22.3.)

**Comment by Jeroen-GroeneBij** [comment]
Issue is still valid in 3.40.8
I tested using the zip file from agiudiceandrea. It does indeed contain one feature with an x and y coordinate well beyond the CRS extent. The extent of this crs is 119.99, 20.41, 122.06, 26.72
The values for x and y are 100 times this extent (and negative). The have both 309 digits (an no decimal point) in stead of an expected two or three before a decimal point.
The issue here is not with a NULL geometry, it is with a geometry that has huge coordinate values.
If you remove the geometry from that single point, the Topology Checker > Must not have duplicates runs flawless.

I'll change the topic title.
